### PR TITLE
Add method to initialize FlexFlow runtime

### PR DESCRIPTION
**Pull request by goliaro**
**Description of changes:**

- [x] Native support for Jupyter and interactive Python, with control of runtime parameters such as the number of GPUs
- [x] Be able to import the flexflow module without immediately starting the runtime
- [x] Make it easier for the user to specify runtime parameters, without needing to type long list of command line arguments
- [x] Debug and cleanup
- [ ] Update docs

The main change introduced by this PR is the `python/flexflow/core/__init__.py` file. This file is run every time one imports the `flexflow.core` package. Before, the file would immediately launch the Legion runtime, setting the number of GPUs to 1 and leaving all other settings (memory per GPU, etc...) unspecified if the corresponding CLI flags were not passed when launching `python`, `flexflow_python` or Jupyter. Now, one can import: `from flexflow.core import *` or `import flexflow.core as ff` without automatically starting the runtime. 

This is useful when testing (we can check that the flexflow package can be imported successfully without needing a GPU), for generating the Python API docs automatically (readthedocs needs to import flexflow to do that), and using any of the utils or functions / utils / objects from the flexflow package without needing a GPU. Before, these things required exporting specific envs, such as `CPU_ONLY_TEST` or `FF_BUILD_DOCS`. Finally, and more importantly, we can now support interactive Python and Jupyter natively, and the user can configure the runtime without needing to pass a long string of command line arguments, which can be more difficult to remember. 

Example of how to run FlexFlow in Jupyter:

```python
import flexflow.core as ff

ff.init_flexflow_runtime(
        num_gpus=4,
        memory_per_gpu=14000,
        zero_copy_memory_per_node=30000,
        only_data_parallel=True
    )
```

Note that when using FlexFlow Serve, the method to initialize the runtime is instead the following (which will also call  `init_flexflow_runtime` after checking the inference arguments):

```python
import flexflow.serve as ff

ff.init(
        num_gpus=4,
        memory_per_gpu=14000,
        zero_copy_memory_per_node=30000,
        tensor_parallelism_degree=4,
        pipeline_parallelism_degree=1
    )
```

**Related Issues:**

Linked Issues:
- Issue #

Issues closed by this PR:
- Closes #

### CODE REVIEW

**Review by jiazhihao — APPROVED** [review]

**Review by xinhaoc — APPROVED** [review]

**Comment by goliaro** [comment]
A couple of additional fixes will be needed to continue supporting flexflow_python effectively, so I'm reverting to draft mode for now. I should be able to push the fixes tomorrow. Please don't merge it until CI passes 😁

File: examples/python/keras/callback.py [hunk 1]
```diff
@@ -66,6 +67,24 @@ def top_level_task():
 
   model.fit(x_train, y_train, epochs=80, callbacks=[mylr_scheduler, VerifyMetrics(ModelAccuracy.CIFAR10_CNN), EpochVerifyMetrics(ModelAccuracy.CIFAR10_CNN)])
 
+def get_configs():
```
Comment (eddy16112) [inline examples/python/keras/callback.py:70]: This function is duplicated in all tests, maybe we can move it into a common file

File: python/flexflow/core/flexflow_cffi.py [hunk 2]
```diff
@@ -15,19 +15,20 @@
...
+from .flexflowlib import ffi, flexflow_library
 
-_FF_BUILD_DOCS = bool(os.environ.get('READTHEDOCS') or os.environ.get("FF_BUILD_DOCS"))
-if not _FF_BUILD_DOCS:
-  from .flexflowlib import ffi, flexflow_library
+def ffc():
+  if not flexflow_already_initialized():
+    raise RuntimeError("Cannot use FlexFlow library before initializing FlexFlow")
   ffc = flexflow_library.lib
+  if ffc is None:
```
Comment (eddy16112) [inline python/flexflow/core/flexflow_cffi.py:29]: why ffc could be None?

Comment (goliaro) [reply python/flexflow/core/flexflow_cffi.py:29]: It shouldn't be None, this is just to catch any potential bug where flexflow is marked as initialized, but the library was not actually initialized in time for the first command that uses it

File: examples/python/keras/callback.py (hunk 1 above)
Comment (goliaro) [reply examples/python/keras/callback.py:70]: good idea!

**Pull request merged**